### ISSUE TITLE
False positive `used-before-assignment` in pattern matching with a guard.

### ISSUE THREAD

**Issue by pgoy-dod** [PyCQA/pylint#5327]
### Bug description

When using pattern matching with a guard, `pylint` claims that the variable the guard uses is used before assignment. 

### Example:

```python
match ("example", "one"):
    case (x, y) if x == "example":  # <-- pylint complains about 'x' here
        print("woo!")
    case _:
        print("boo...")
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
************* Module a
a.py:2:19: E0601: Using variable 'x' before assignment (used-before-assignment)

-------------------------------------
Your code has been rated at -10.00/10
```

### Expected behavior

Pattern matching with guards is correctly parsed and supported.

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.4
Python 3.10.0 (default, Oct 13 2021, 06:45:00) [Clang 13.0.0 (clang-1300.0.29.3)]
```

### OS / Environment

Mac OSX Big Sur 11.6 (20G165)

**Comment by pilatte** [comment]
This bug is still present in:
```
pylint 2.12.2
astroid 2.9.3
Python 3.10.2 (main, Feb  1 2022, 16:44:51) [Clang 12.0.5 (clang-1205.0.22.9)]
```
Here my example code
```py
import sys
match sys.argv[1:]:         # vvvvvvvvv Using variable 'direction' before assignment
    case ["go", direction] if direction in ["up", "down", "left", "right"]:
        print(f"going {direction}")
    case command:
        print(f"unknown command: {' '.join(command)}")
```

**Comment by jborman-exos** [comment]
Also hitting this false-positive, is there a plan to include this in an upcoming release?

**Comment by spagh-eddie** [comment]
I too suffer from this issue

**Comment by mthuurne** [comment]
I'm running into the same issue using:
```
pylint 2.14.4
astroid 2.11.6
Python 3.10.4 (main, Mar 26 2022, 21:00:27) [GCC]
```

My testcase:
```py
def describe(i: int) -> str:
    match i:
        case n if n % 2 == 0:
            return "even"
        case _:
            return "odd"
```

**Comment by maujim** [comment]
If the same variable is used as a match guard twice, pylint won't complain the second time. 

```python class:lineNo
l = [7, 7]
match l:
    case [_, n] if n >= 4: 		# E0601: Using variable 'n' before assignment
        print("hi")

t = (4, "abc")
match t:
    case (n, _) if n < 12:		# not triggered here
        print("hi")
```

If the line `case (n, _) if n < 12:` is changed to use `x` instead of `n`, the false positive triggers again.

version:
